File: jarvis-personal/native/ios/DincrKit/Sources/DincrCore/APIClient.swift
import Foundation

/// Supplies the bearer token for API calls. `forceRefresh` is used once after a 401.
public protocol AccessTokenProvider: Sendable {
    func accessToken(forceRefresh: Bool) async throws -> String
}

/// Abstracts the network so tests can script responses.
public protocol HTTPTransport: Sendable {
    func send(_ request: URLRequest) async throws -> (Data, HTTPURLResponse)
}

public struct URLSessionTransport: HTTPTransport {
    let session: URLSession
    public init(session: URLSession = .shared) { self.session = session }

    public func send(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        return (data, http)
    }
}

/// DINCR API client. Same contract as `frontend/src/lib/authenticatedFetch.js`:
/// - `Authorization: Bearer`, `Accept-Language`, `X-Request-ID` (stable across retries),
///   `X-Retry-Attempt`;
/// - safe methods (GET/HEAD) retry up to 2 times on 408, 425, 429, 502, 503, 504 and on
///   network loss; writes are never retried automatically, and creates carry an
///   `X-Idempotency-Key` so a repeated submit is answered from the first one;
/// - a 401 refreshes the session once and repeats the request;
/// - 20 s timeout per attempt.
public struct APIClient: Sendable {
    public static let retryableStatus: Set<Int> = [408, 425, 429, 502, 503, 504]

    let baseURL: URL
    let tokens: AccessTokenProvider
    let transport: HTTPTransport
    let language: AppLanguage
    let timeout: TimeInterval
    let backoff: @Sendable (Int) async -> Void

    public init(
        baseURL: URL, tokens: AccessTokenProvider, transport: HTTPTransport = URLSessionTransport(),
        language: AppLanguage = .current, timeout: TimeInterval = 20,
        backoff: @escaping @Sendable (Int) async -> Void = { attempt in
            try? await Task.sleep(for: .milliseconds(400 * (1 << attempt)))
        }
    ) {
        self.baseURL = baseURL; self.tokens = tokens; self.transport = transport
        self.language = language; self.timeout = timeout; self.backoff = backoff
    }

    static let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return decoder
    }()

    static let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.keyEncodingStrategy = .convertToSnakeCase
        return encoder
    }()

    public func get<Response: Decodable>(_ path: String, query: [URLQueryItem] = [], as type: Response.Type = Response.self) async throws -> Response {
        try await perform(method: "GET", path: path, query: query, body: nil)
    }

    public func send<Body: Encodable, Response: Decodable>(_ method: String, _ path: String, body: Body, idempotencyKey: String? = nil, as type: Response.Type = Response.self) async throws -> Response {
        let data: Data
        do { data = try Self.encoder.encode(body) } catch { throw APIError.decoding(language) }
        return try await perform(method: method, path: path, query: [], body: data, idempotencyKey: idempotencyKey)
    }

    public func send<Response: Decodable>(_ method: String, _ path: String, as type: Response.Type = Response.self) async throws -> Response {
        try await perform(method: method, path: path, query: [], body: nil)
    }

    func perform<Response: Decodable>(method: String, path: String, query: [URLQueryItem], body: Data?, idempotencyKey: String? = nil) async throws -> Response {
        let requestID = UUID().uuidString.lowercased()
        let safe = method == "GET" || method == "HEAD"
        let maxRetries = safe ? 2 : 0
        var attempt = 0
        var refreshed = false
        var token = try await tokens.accessToken(forceRefresh: false)

        while true {
            var request = URLRequest(url: url(path, query: query), timeoutInterval: timeout)
            request.httpMethod = method
            request.httpBody = body
            request.setValue("application/json", forHTTPHeaderField: "Accept")
            if body != nil { request.setValue("application/json", forHTTPHeaderField: "Content-Type") }
            request.setValue(language.rawValue, forHTTPHeaderField: "Accept-Language")
            request.setValue(requestID, forHTTPHeaderField: "X-Request-ID")
            request.setValue(String(attempt), forHTTPHeaderField: "X-Retry-Attempt")
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            if let idempotencyKey { request.setValue(idempotencyKey, forHTTPHeaderField: "X-Idempotency-Key") }

            let data: Data
            let response: HTTPURLResponse
            do {
                (data, response) = try await transport.send(request)
            } catch let error as URLError {
                // A cancelled task (the screen went away) is not an outage: never show "offline".
                if error.code == .cancelled { throw CancellationError() }
                if attempt < maxRetries, Self.isTransient(error) {
                    await backoff(attempt); attempt += 1; continue
                }
                throw error.code == .timedOut ? APIError.timeout(language) : APIError.offline(language)
            }

            if response.statusCode == 401, !refreshed {
                refreshed = true
                token = try await tokens.accessToken(forceRefresh: true)
                continue
            }
            if Self.retryableStatus.contains(response.statusCode), attempt < maxRetries {
                await backoff(attempt); attempt += 1; continue
            }
            guard (200..<300).contains(response.statusCode) else {
                throw APIError.from(status: response.statusCode, body: data, language: language, requestID: requestID)
            }
            do {
                return try Self.decoder.decode(Response.self, from: data.isEmpty ? Data("{}".utf8) : data)
            } catch {
                throw APIError.decoding(language)
            }
        }
    }

    func url(_ path: String, query: [URLQueryItem]) -> URL {
        var components = URLComponents(url: baseURL.appending(path: path), resolvingAgainstBaseURL: false)!
        if !query.isEmpty { components.queryItems = query }
        return components.url!
    }

    static func isTransient(_ error: URLError) -> Bool {
        [.notConnectedToInternet, .networkConnectionLost, .timedOut, .cannotConnectToHost, .dnsLookupFailed].contains(error.code)
    }
}

File: jarvis-personal/native/android/core/data/src/main/kotlin/com/dincr/data/ApiClient.kt
package com.dincr.data

import java.io.IOException
import java.io.InterruptedIOException
import java.util.UUID
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

/** Supplies the bearer token. `forceRefresh` is used once after a 401. */
fun interface AccessTokenProvider {
    suspend fun accessToken(forceRefresh: Boolean): String
}

/** A plain HTTP exchange, so tests can script responses without a server. */
data class HttpRequest(val method: String, val url: String, val headers: Map<String, String>, val body: String?)
data class HttpResponse(val status: Int, val body: String)

fun interface HttpTransport {
    /** Throws [IOException] on network failure. */
    suspend fun send(request: HttpRequest): HttpResponse
}

class OkHttpTransport(
    private val client: OkHttpClient = OkHttpClient.Builder()
        .callTimeout(20, TimeUnit.SECONDS)
        .build(),
) : HttpTransport {
    override suspend fun send(request: HttpRequest): HttpResponse = withContext(Dispatchers.IO) {
        val body = request.body?.toRequestBody("application/json".toMediaType())
            ?: if (request.method == "GET" || request.method == "HEAD") null else ByteArray(0).toRequestBody(null)
        val okRequest = Request.Builder().url(request.url).method(request.method, body)
            .apply { request.headers.forEach { (name, value) -> header(name, value) } }
            .build()
        client.newCall(okRequest).execute().use { response -> HttpResponse(response.code, response.body?.string().orEmpty()) }
    }
}

/**
 * DINCR API client. Same contract as `frontend/src/lib/authenticatedFetch.js` and iOS `APIClient`:
 * Bearer token, Accept-Language, stable X-Request-ID, X-Retry-Attempt; safe methods retry up to
 * 2 times on 408/425/429/502/503/504 or network loss; writes are never retried and creates carry
 * an `X-Idempotency-Key`; a 401 refreshes the session once.
 */
class ApiClient(
    private val baseUrl: String,
    private val tokens: AccessTokenProvider,
    private val transport: HttpTransport = OkHttpTransport(),
    private val language: AppLanguage = AppLanguage.current(),
    private val backoff: suspend (Int) -> Unit = { attempt -> delay(400L shl attempt) },
) {
    val json = Json { ignoreUnknownKeys = true; explicitNulls = false; encodeDefaults = true }

    suspend inline fun <reified T> get(path: String, query: Map<String, String> = emptyMap()): T =
        decode(perform("GET", path, query, null))

    suspend inline fun <reified B, reified T> send(method: String, path: String, body: B, idempotencyKey: String? = null): T =
        decode(perform(method, path, emptyMap(), json.encodeToString(body), idempotencyKey))

    suspend inline fun <reified T> send(method: String, path: String): T =
        decode(perform(method, path, emptyMap(), null))

    inline fun <reified T> decode(body: String): T =
        try {
            json.decodeFromString<T>(body.ifEmpty { "{}" })
        } catch (error: Exception) {
            throw ApiError.decoding(currentLanguage)
        }

    val currentLanguage: AppLanguage get() = language

    suspend fun perform(method: String, path: String, query: Map<String, String>, body: String?, idempotencyKey: String? = null): String {
        val requestId = UUID.randomUUID().toString()
        val safe = method == "GET" || method == "HEAD"
        val maxRetries = if (safe) 2 else 0
        var attempt = 0
        var refreshed = false
        var token = tokens.accessToken(false)
        val url = baseUrl.trimEnd('/').plus(path).toHttpUrl().newBuilder()
            .apply { query.forEach { (key, value) -> addQueryParameter(key, value) } }
            .build().toString()

        while (true) {
            val headers = buildMap {
                put("Accept", "application/json")
                if (body != null) put("Content-Type", "application/json")
                put("Accept-Language", language.tag)
                put("X-Request-ID", requestId)
                put("X-Retry-Attempt", attempt.toString())
                put("Authorization", "Bearer $token")
                if (idempotencyKey != null) put("X-Idempotency-Key", idempotencyKey)
            }
            val response = try {
                transport.send(HttpRequest(method, url, headers, body))
            } catch (error: IOException) {
                if (attempt < maxRetries) { backoff(attempt); attempt += 1; continue }
                throw if (error is InterruptedIOException) ApiError.timeout(language) else ApiError.offline(language)
            }
            if (response.status == 401 && !refreshed) {
                refreshed = true
                token = tokens.accessToken(true)
                continue
            }
            if (response.status in RETRYABLE_STATUS && attempt < maxRetries) {
                backoff(attempt); attempt += 1; continue
            }
            if (response.status !in 200..299) throw ApiError.from(response.status, response.body, language, requestId)
            return response.body
        }
    }

    companion object {
        val RETRYABLE_STATUS = setOf(408, 425, 429, 502, 503, 504)
    }
}
